### PR TITLE
Implement DDR findmodules command

### PR DESCRIPTION
**Pull request by acrowthe**
Implement DDR findmodules command

Added new command: !findmodules. Display all modules matching one of
the below options:
- all (default). Lists all modules loaded by the runtime
- name <module name>. Lists all modules with the same name as the parameter
- requires <module name>. Lists all modules that require the input module
- package <package name>. Lists the loaded module that owns the input package

Implements part of the new proposal within #1859

Created ModularityHelper.java
- Will assist in implementing other parts of the second proposal,
such as !dumppackage and !dumpmodule
- Used to traverse module and package lists
- Provides filter and output support

Example modules:
moduleA
```
module moduleA {
	exports moduleA.packageA;
	requires moduleB;
}
```
moduleB
```
module moduleB {
	exports moduleB.packageB to moduleA;
}
```
Example calls
!findmodules name moduleA
```
moduleA                        !j9module 0x00007FAC203C55B8
```
!findmodules requires moduleB
```
moduleA                        !j9module 0x00007FCA242F0960
```
!findmodules package moduleA.packageA
```
moduleA                        !j9module 0x00007FAC203C55B8
```
!findmodules
or
!findmodules all
```
jdk.naming.rmi                 !j9module 0x00007FAC20335E48
java.desktop                   !j9module 0x00007FAC203A1CB8
java.xml                       !j9module 0x00007FAC203B4D10
java.security.sasl             !j9module 0x00007FAC20335C98
jdk.management.jfr             !j9module 0x00007FAC203CD958
java.logging                   !j9module 0x00007FAC20335A58
java.datatransfer              !j9module 0x00007FAC20399068
java.naming                    !j9module 0x00007FAC20399410
jdk.jfr                        !j9module 0x00007FAC203A1490
java.rmi                       !j9module 0x00007FAC202EECC8
java.prefs                     !j9module 0x00007FAC20335788
openj9.sharedclasses           !j9module 0x00007FAC203358F0
java.management                !j9module 0x00007FAC20336118
java.management.rmi            !j9module 0x00007FAC203A1B08
jdk.management                 !j9module 0x00007FAC202EF148
java.base                      !j9module 0x00007FAC2008EAC8
java.compiler                  !j9module 0x00007FAC203998D8
java.security.jgss             !j9module 0x00007FAC203C5690
jdk.localedata                 !j9module 0x00007FAC202EF6E8
jdk.security.auth              !j9module 0x00007FAC202EEAD0
jdk.charsets                   !j9module 0x00007FAC202EEC38
jdk.crypto.ec                  !j9module 0x00007FAC203359C8
jdk.security.jgss              !j9module 0x00007FAC20335F20
jdk.crypto.cryptoki            !j9module 0x00007FAC203A13B8
java.xml.crypto                !j9module 0x00007FAC20390718
jdk.zipfs                      !j9module 0x00007FAC203A1C28
java.smartcardio               !j9module 0x00007FAC20335818
jdk.naming.dns                 !j9module 0x00007FAC20399188
moduleA                        !j9module 0x00007FAC203C55B8
jdk.javadoc                    !j9module 0x00007FAC203CD2E0
moduleC                        !j9module 0x00007FAC20399020
jdk.jartool                    !j9module 0x00007FAC20399260
openj9.gpu                     !j9module 0x00007FAC20335BC0
jdk.unsupported.desktop        !j9module 0x00007FAC203CDA30
moduleB                        !j9module 0x00007FAC203CDB08
jdk.jdeps                      !j9module 0x00007FAC203A1880
jdk.internal.opt               !j9module 0x00007FAC20335FF8
openj9.cuda                    !j9module 0x00007FAC2038FCF8
jdk.compiler                   !j9module 0x00007FAC2038FDD0
jdk.jlink                      !j9module 0x00007FAC202EF3D0
```

Signed-off-by: Andrew Crowther <acrowthe3388@gmail.com>

### CODE REVIEW

**Comment by acrowthe** [comment]
@tajila Let me know if the description change is valid. All other concerns have been addressed

**Comment by tajila** [comment]
@acrowthe I mean't please show the output of some of your commands (e.g. !findmodule all ...)

**Comment by tajila** [comment]
```
!findmodules requires moduleA

java.base                      !j9module 0x00007FAC2008EAC8
moduleC                        !j9module 0x00007FAC20399020
moduleB                        !j9module 0x00007FAC203CDB08
```

It looks like that is doing the opposite o what it should do. 

`!findmodules requires moduleA` should print all the modules that require `moduleA`

File: debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/commands/FindModulesCommand.java [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+	public void run(String command, String[] args, Context context, PrintStream out) throws DDRInteractiveCommandException 
+	{
+		Status status = Status.HELP;
+		String filterArg = null;
+		ModuleIteratorFilter moduleFilter = null;
+		PackageIteratorFilter packageFilter = null;
+		switch (args.length) {
+		case 0:
+			moduleFilter = ModularityHelper::moduleFilterMatchAll;
+			status = Status.MODULE;
```
Comment (tajila) [inline debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/commands/FindModulesCommand.java]: You can remove the `Status.` and just have MODULE

Comment (acrowthe) [reply debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/commands/FindModulesCommand.java]: It dows not compile when I leave out Status.

**Comment by acrowthe** [comment]
@tajila I have fixed !findmodules require, and updated the example call to match

File: debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/commands/FindModulesCommand.java (hunk 1 above)
Comment (tajila) [reply debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/commands/FindModulesCommand.java]: `import static ...Status.HELP` 
...

**Comment by tajila** [comment]
@keithc-ca  Can you please review this PR

File: debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/ModularityHelper.java [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+				while (slotIterator.hasNext()) {
+					J9PackagePointer packagePtr = slotIterator.next();
+					if (filter.filter(packagePtr, filterArg)) {
+						outputter.print(packagePtr, out);
+					}
+				}
+			}
+		}
+	}
+}
```
Comment (keithc-ca) [inline debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/ModularityHelper.java]: Please don't use internal tabs (those other than the group at the beginning of a line).
All source files should end with a final newline.

Comment (acrowthe) [reply debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/ModularityHelper.java]: All internal tabs within `ModularityHelper.java` and `FindModulesCommand.java` have been replaced by spaces. Missing end newline has been added.

**Comment by acrowthe** [comment]
@keithc-ca I believe I have fixed your concerns.

**Comment by keithc-ca** [comment]
@acrowthe Have you arranged for the relevant documentation to be updated?

**Comment by acrowthe** [comment]
> @acrowthe Have you arranged for the relevant documentation to be updated?

I have not made arrangements, though once all parts of the proposal in #1859 have been implemented I will make sure that that occurs. If you would like for the documentation to be updated with each part of #1859, I could also do that.

**Comment by acrowthe** [comment]
@keithc-ca Documentation for this command has been submitted to #3102.

**Comment by acrowthe** [comment]
Fixed slight misformat in the addCommand string: Changed 
```
" [all|name <moduleName>|requires <moduleName>|package <packageName>] | help"
```
to
```
" [all|name <moduleName>|requires <moduleName>|package <packageName>] | help"
```

File: debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/commands/FindModulesCommand.java [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+			}
+		}
+		return result;
+	}
+
+	private static boolean filterPackageName(J9PackagePointer packagePtr, String targetName) throws CorruptDataException  {
+		return J9UTF8Helper.stringValue(packagePtr.packageName()).equals(targetName);
+	}
+
+	void printHelp(PrintStream out) {
```
Comment (keithc-ca) [inline debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/commands/FindModulesCommand.java]: This method can be `private static`.

**Comment by acrowthe** [comment]
@keithc-ca Changes made and manually tested.

The square brackets around the description were due to `!findmodules` being a valid command without any further arguments: The brackets have been adjusted to better match this use case

File: debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/commands/FindModulesCommand.java [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+			break;
+		case 1:
+			if (args[0].equalsIgnoreCase("all")) {
+				moduleFilter = ModularityHelper::moduleFilterMatchAll;
+				status = MODULE;
+			} /* else stay status.HELP */
+			break;
+		case 2:
+			filterArg = args[1]; /* use args[1] as a hex address */
+			switch(args[0]) {
```
Comment (keithc-ca) [inline debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/commands/FindModulesCommand.java]: nit: Inconsistent formatting (there should be a space after `switch`, like line 71).

**Comment by keithc-ca** [comment]
jenkins sanity plinux jdk10

**Comment by keithc-ca** [comment]
jenkins test sanity plinux jdk10

**Comment by keithc-ca** [comment]
jenkins test sanity plinux jdk10

**Comment by keithc-ca** [comment]
jenkins copyright check

**Comment by keithc-ca** [comment]
jenkins copyright check

**Comment by keithc-ca** [comment]
jenkins test sanity plinux jdk10

**Comment by keithc-ca** [comment]
jenkins test sanity zlinux jdk10

**Comment by keithc-ca** [comment]
I believe the plinux failure was an infrastructure issue. The code is not platform-specific and the zlinux build was successful - merging.

**Pull request merged**